tsl/src/data_node.c

```diff
@@ -574,10 +575,13 @@ connect_for_bootstrapping(const char *node_name, const char *const host, int32 p
 static void
 data_node_validate_extension_availability(TSConnection *conn)
 {
-	PGresult *res = remote_connection_execf(conn,
-											"SELECT default_version, installed_version FROM "
-											"pg_available_extensions WHERE name = %s",
-											quote_literal_cstr(EXTENSION_NAME));
+	bool compatible;
+
+	PGresult *res =
+		remote_connection_execf(conn,
+								"SELECT version FROM pg_available_extension_versions WHERE name = "
+								"%s AND version ~ '\\d+.\\d+.\\d+.*' ORDER BY version DESC",
```

Comment:
Why is the version regexp needed here? Do we expect versions with a different format and why should those not be included in that case?

It is also unclear to me which use case this change addresses. I guess it is trying to check that there is at least one compatible version available for installation?

But there seems to be many different cases to handle here:

1. The database does not exist on the data node. Requirement: there must be a compatible extension version available
2. The database exists but does not have the extension installed. Same requirement as 1.
3. The database exists and extension is installed. Requirement: the installed extension must be a compatible version.


Which of these cases do we handle? It seems this change mostly targets case 1 and 2 above.

Reply:
Regexp is there to avoid tests breaking - since some of our so mocks don't follow our semantic versioning. If we would change the way we do versioning then that will require changes in few other places in our code (eg. logic that does version compatibility checking).

Right, your observations are pretty much correct. Let me explain a bit more. We have two types of checks: 1. check available versions 2. check installed version. Check 1 is what I was focused on here and that check if performed only
when trying to bootstrap DN. Check 2 is performed later on after we bootstrap DN and also each time we establish a connection.
This means that with this change in place the bootstrap process will not fail if it finds some available extension but can still fail later on if installed version is incompatible with version on AN. This would mean that user will need to do ALTER
EXTENSION UPDATE on problematic DN and then try adding the DN again. Hope this makes it more clear now.

Reply:
Is this same function used for case 2? Do we support case 3 (and which function is used to validate the extension in that case)?

Would be good to make this clear in comments or the function description. For instance, it would be good to describe exactly how the function is used (i.e., which cases above).

Reply:
Well, this function is only used when bootstrapping parameter is true (that's by default). So it will run for all three case you've mentioned. No matter if DB/extension exists or not.